File: algorithms/3_hard/B/00123_best_time_to_buy_and_sell_stock_iii/state_machine.c
#define K 2

typedef ptrdiff_t isize;

static inline int max_int(const int a, const int b)
{
    return a > b ? a : b;
}

int maxProfit(const int* prices, const int len_i32)
{
    assert(len_i32 > 0 && prices != NULL);
    const isize len = len_i32;
    int dp[K + 1] = { 0 };
    int hold[K + 1];
    for (isize i = 0; i < K + 1; i += 1) { hold[i] = INT_MIN; }

    for (isize i = 0; i < len; i += 1) {
        const int p = prices[i];

        for (isize j = K; j > 0; j -= 1) {
            dp[j] = max_int(dp[j], hold[j] + p);
            hold[j] = max_int(hold[j], dp[j - 1] - p);
        }
    }

    return dp[K];
}

File: algorithms/3_hard/B/00123_best_time_to_buy_and_sell_stock_iii/state_machine.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        const k: usize = 2;
        let mut hold = [i32::MIN; k + 1];
        let mut dp = [0_i32; k + 1];

        for p in prices {
            for i in (1..=k).rev() {
                dp[i] = dp[i].max(hold[i] + p);
                hold[i] = hold[i].max(dp[i - 1] - p);
            }
        }

        dp[k]
    }
}
